lib/elixir/lib/partition_supervisor.ex

```diff
@@ -202,32 +202,63 @@ defmodule PartitionSupervisor do
   def start_child(mod, fun, args, name, partition) do
     case apply(mod, fun, args) do
       {:ok, pid} ->
-        :ets.insert(name, {partition, pid})
+        register_child(name, {partition, pid})
         {:ok, pid}
 
       {:ok, pid, info} ->
-        :ets.insert(name, {partition, pid})
+        register_child(name, {partition, pid})
         {:ok, pid, info}
 
       other ->
         other
     end
   end
 
+  defp register_child(name, key) when is_atom(name) do
+    :ets.insert(name, key)
+  end
+
+  defp register_child(name, key) when is_tuple(name) do
+    Registry.register(@registry, name, key)
+  end
+
   @impl true
   def init({name, partitions, children, init_opts}) do
+    init_partitions(name, partitions)
+
+    Supervisor.init(children, Keyword.put_new(init_opts, :strategy, :one_for_one))
+  end
+
+  defp init_partitions(name, partitions) when is_atom(name) do
     :ets.new(name, [:set, :named_table, :protected, read_concurrency: true])
     :ets.insert(name, {:partitions, partitions})
-    Supervisor.init(children, Keyword.put_new(init_opts, :strategy, :one_for_one))
+  end
+
+  defp init_partitions(name, partitions) when is_tuple(name) do
+    child_spec = {Registry, keys: :duplicate, name: @registry}
+
+    case Supervisor.start_child(:elixir_sup, child_spec) do
+      {:ok, _pid} ->
+        Registry.register(@registry, name, {:partitions, partitions})
```

Comment:
I think the key must be `self()`. That's because we can have different via tuples resolving to the same name.